### ISSUE TITLE
[BUG] OpenStack node driver fails to delete node if cacert is provided

### ISSUE THREAD

**Issue by fstarkenberg** [rancher/rancher#39222]
**Rancher Server Setup**
- Rancher version: v2.6.8
- Installation option (Docker install/Helm Chart): Helm Chart
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): RKE2
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: v1.24.4+rke2r1
- Cluster Type (Local/Downstream): Local cluster using Openstack node driver to provision downstream RKE2 cluster
- If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

**Describe the bug**
When provisioning a new RKE2 cluster using the OpenStack node driver there is a field for cacert (Description: "File contents for cacert"). The resulting openstackconfig object looks something like this:
```yaml
kubectl get openstackconfig -o yaml
apiVersion: v1
items:
- activeTimeout: "200"
  apiVersion: rke-machine-config.cattle.io/v1
  applicationCredentialId: <censored>
  applicationCredentialName: ""
  applicationCredentialSecret: <censored>
  authUrl: https://<censored>:5000/v3
  availabilityZone: ""
  bootFromVolume: true
  cacert: |
    -----BEGIN CERTIFICATE-----
    MIIG0zCCBbugAwIBAgITHQA3ejXwrpZ/eakFogAAADd6NTANBgkqhkiG9w0BAQsF
    ...
```

When provisioning nodes the rancher-machine job creates a pod with the following args:
```yaml
  - args:
    - "--driver-download-url="
    - "--driver-hash="
    - "--secret-namespace=fleet-<censored>-dev"
    - "--secret-name=<censored>-pool1-2b1af3a2-bbz5f-machine-state"
    - create
    - "--driver=openstack"
    - "--custom-install-script=/run/secrets/machine/value"
    - "--openstack-active-timeout=200"
    - "--openstack-application-credential-id=<censored>"
    - "--openstack-application-credential-secret=<censored>"
    - "--openstack-auth-url=https://<censored>:5000/v3"
    - "--openstack-boot-from-volume"
    - "--openstack-cacert=/path/to/machine/files/cacert"
    ...
    - mountPath: "/path/to/machine/files"
      name: machine-files
      readOnly: true
    ...
  volumes:
  - name: machine-files
    secret:
      defaultMode: 420
      items:
      - key: cacert
        path: cacert
      - key: userDataFile
        path: userDataFile
      secretName: <censored>-pool1-2b1af3a2-bbz5f-machine-provision
  ...
```
So far so good, everything works as expected. The problem arise when deleting a node.
The job/pod that is scheduled to delete the node does not contain any `machine-files` mount at all. Which leads to the following error
```yaml
kubectl logs -f <censored>-pool1-aab7fec4-97l49-machine-provision--1-7nfpn
About to remove <censored>-pool1-aab7fec4-97l49
WARNING: This action will delete both local reference and remote instance.
(<censored>-pool1-aab7fec4-97l49) deleting instance... map[MachineId:b9425915-0eb7-4618-b3ac-ec47f3963251]
(<censored>-pool1-aab7fec4-97l49) Deleting OpenStack instance...
(<censored>-pool1-aab7fec4-97l49) Authenticating... map[ApplicationCredentialId:<censored>
ApplicationCredentialName: AuthUrl:https://<censored>:5000/v3
CaCert:/path/to/machine/files/cacert
DomainId: DomainName:Default Insecure:false TenantDomainID: TenantDomainName: TenantID:4d3100a25853466a9f914e0535aa418e TenantName: UserDomainID: UserDomainName: UserId: Username:]
Error removing host "<censored>-pool1-aab7fec4-97l49": open /path/to/machine/files/cacert: no such file or directory
```

Basically the `machines-files` secret is still needed when deleting a node but seems to have been lost somewhere along the way.

**To Reproduce**
Use the OpenStack node driver to provision a cluster with `cacert` field in OpenstackConfig filled in.

**Result**
`open /path/to/machine/files/cacert: no such file or directory`

**Expected Result**
The certificate in `cacert` should be mounted to ` /path/to/machine/files/cacert` and the node should be deleted.

**Additional context**
Happy to clarify if something is unclear.

**Comment by thomasferrandiz** [comment]
I encountered the same issue while doing tests on our internal Openstack for a customer.
I propose this fix which works fine in my tests: 
https://github.com/thomasferrandiz/rancher/commit/287df252595024d62c753071afbfba8903425f83

Basically it just runs the `constructFilesSecret` function when deleting the node to add the required secrets to the rancher-machine pod.

**Comment by Oats87** [comment]
Thanks for the comprehensive issue detailing the problem and providing a potential solution.

I don't necessarily see any immediate issues with running `constructFilesSecret` as you tested in your fix on both creation and deletion; however, I would also like to state that from what I can tell, this appears to be an issue specific to OpenStack (`constructFilesSecret` requires the `SchemaToDriverFields` to be populated for the driver, and `cacert` is only defined for `openstack` here: https://github.com/rancher/rancher/blob/7913e283ff90eebd1f8cc48860a6b410d72cd4dd/pkg/controllers/management/node/controller.go#L62

Given the fact that we do not officially support `OpenStack` as a machine driver, I will need to defer to @Sahota1225 and @snasovich on our ability to resolve this for `v2.7.2`.

**Comment by snasovich** [comment]
@Oats87 , thank you for reviewing this.

@thomasferrandiz , at this time 2.7.2 is past all kinds of "freeze" dates so it's too late to add this change, especially since there are some regression concerns as the changes are in the common machine provisioning controller.

Though FWIW at a glance it looks like this is an old regression caused by https://github.com/rancher/rancher/pull/36110 (which by itself is a fix to regression from refactoring - https://github.com/rancher/rancher/pull/35925).

Still, it's too late to include it in 2.7.2 as we would want more regression testing done. Adding to Q2 release queue.

**Comment by snasovich** [comment]
@thomasferrandiz , the PR got merged - sorry for the delay. There is also a new `rancher/rancher:v2.7-head` image with the changes in, so please have someone on your team validate it.
cc: @kkaempf @pgonin

**Comment by thomasferrandiz** [comment]
@snasovich 
I tested the creation and deletion of a node with image `rancher/rancher:v2.7-head` and everything went well.
The node is deleted both from rancher and from openstack.
